File: controller/nginx.py
"""
Docker-based Nginx management for AutoServe.
Manages Nginx configuration through Docker container operations.
"""

import docker
import logging
import tempfile
import shutil
import os
from jinja2 import Template
from pathlib import Path
from typing import List, Dict
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

logger = logging.getLogger(__name__)

class DockerNginxManager:
    def __init__(self, nginx_container_name: str = None, conf_dir: str = None, template_path: str = None):
        self.docker_client = docker.from_env()
        
        self.nginx_container_name = nginx_container_name or os.getenv("AUTOSERVE_NGINX_CONTAINER")
        if not self.nginx_container_name:
            logger.error("AUTOSERVE_NGINX_CONTAINER environment variable is required. Please set it in .env file.")
            raise RuntimeError("Missing required environment variable: AUTOSERVE_NGINX_CONTAINER")
            
        self.conf_dir = Path(conf_dir or os.getenv("AUTOSERVE_NGINX_CONF_DIR"))
        if not conf_dir and not os.getenv("AUTOSERVE_NGINX_CONF_DIR"):
            logger.error("AUTOSERVE_NGINX_CONF_DIR environment variable is required. Please set it in .env file.")
            raise RuntimeError("Missing required environment variable: AUTOSERVE_NGINX_CONF_DIR")
        self.template_path = template_path or "docker_configs/nginx_template.conf"
        self._load_template()
        
        # Ensure config directory exists
        self.conf_dir.mkdir(parents=True, exist_ok=True)
        
        # Ensure nginx container is running
        self._ensure_nginx_container()
        
    def _load_template(self):
        """Load the Nginx configuration template."""
        try:
            if Path(self.template_path).exists():
                template_content = Path(self.template_path).read_text()
            else:
                logger.warning(f"Template file {self.template_path} not found")
            
            self.template = Template(template_content)
        except Exception as e:
            logger.error(f"Failed to load template: {e}")

    def _ensure_nginx_container(self):
        """Ensure the nginx container is running."""
        try:
            container = self.docker_client.containers.get(self.nginx_container_name)
            if container.status != "running":
                logger.info(f"Starting nginx container {self.nginx_container_name}")
                container.start()
            logger.info(f"Nginx container {self.nginx_container_name} is running")
        except docker.errors.NotFound:
            logger.error(f"Nginx container {self.nginx_container_name} not found. Please start the AutoServe infrastructure.")
            raise Exception(f"Nginx container {self.nginx_container_name} not found")
        except Exception as e:
            logger.error(f"Failed to ensure nginx container: {e}")
            raise

    def _get_nginx_container(self):
        """Get the nginx container object."""
        try:
            return self.docker_client.containers.get(self.nginx_container_name)
        except docker.errors.NotFound:
            raise Exception(f"Nginx container {self.nginx_container_name} not found")

    def update_upstreams(self, app_name: str, servers: List[Dict[str, str]]):
        """Update nginx upstream configuration for an app."""
        try:
            if not servers:
                logger.warning(f"No servers provided for app {app_name}, removing config")
                self.remove_app_config(app_name)
                return
            
            # Render the configuration
            config = self.template.render(app=app_name, servers=servers)
            conf_path = self.conf_dir / f"{app_name}.conf"
            
            # Write configuration to file
            with tempfile.NamedTemporaryFile(mode='w', delete=False, 
                                           dir=self.conf_dir, suffix='.tmp') as tmp_file:
                tmp_file.write(config)
                tmp_path = tmp_file.name
            
            # Test nginx configuration
            nginx_container = self._get_nginx_container()
            test_result = nginx_container.exec_run(
                ["nginx", "-t"]
            )
            
            if test_result.exit_code != 0:
                logger.error(f"Nginx config test failed: {test_result.output}")
                Path(tmp_path).unlink()  # Clean up temp file
                return
            
            # Move temp file to final location
            shutil.move(tmp_path, conf_path)
            
            # Reload nginx
            reload_result = nginx_container.exec_run(
                ["nginx", "-s", "reload"]
            )
            
            if reload_result.exit_code != 0:
                logger.error(f"Nginx reload failed: {reload_result.output}")
                # Try to restore previous config if it existed
                self._restore_previous_config(app_name)
            else:
                logger.info(f"Updated nginx config for {app_name} with {len(servers)} servers")
                
        except Exception as e:
            logger.error(f"Failed to update nginx config for {app_name}: {e}")
    
    def remove_app_config(self, app_name: str):
        """Remove nginx configuration for an app."""
        try:
            conf_path = self.conf_dir / f"{app_name}.conf"
            if conf_path.exists():
                conf_path.unlink()
                
                # Test and reload nginx
                nginx_container = self._get_nginx_container()
                test_result = nginx_container.exec_run(["nginx", "-t"])
                if test_result.exit_code == 0:
                    nginx_container.exec_run(["nginx", "-s", "reload"])
                    logger.info(f"Removed nginx config for {app_name}")
                else:
                    logger.error(f"Nginx config test failed after removing {app_name}")
                    
        except Exception as e:
            logger.error(f"Failed to remove nginx config for {app_name}: {e}")
    
    def _restore_previous_config(self, app_name: str):
        """Attempt to restore a previous working configuration."""
        # This is a placeholder for more sophisticated backup/restore logic
        logger.warning(f"Config restore not implemented for {app_name}")
    
    def get_nginx_status(self) -> Dict:
        """Get nginx status information."""
        try:
            nginx_container = self._get_nginx_container()
            
            # Get nginx status via container exec
            result = nginx_container.exec_run(
                ["curl", "-s", "http://localhost/nginx_status"]
            )
            
            if result.exit_code == 0:
                # Decode bytes to string if needed
                output = result.output
                if isinstance(output, bytes):
                    output = output.decode('utf-8')
                return self._parse_nginx_status(output)
            else:
                # Decode bytes to string if needed for error details
                error_details = result.output
                if isinstance(error_details, bytes):
                    error_details = error_details.decode('utf-8')
                return {"error": "Failed to get nginx status", "details": error_details}
                
        except Exception as e:
            logger.error(f"Failed to get nginx status: {e}")
            return {"error": str(e)}
    
    def _parse_nginx_status(self, status_text: str) -> Dict:
        """Parse nginx stub_status output."""
        try:
            lines = status_text.strip().split('\n')
            
            # Parse active connections
            active_line = lines[0]  # "Active connections: 1"
            active_connections = int(active_line.split(':')[1].strip())
            
            # Parse server stats
            stats_line = lines[2]  # "1 2 3"
            accepts, handled, requests = map(int, stats_line.split())
            
            # Parse reading/writing/waiting
            rw_line = lines[3]  # "Reading: 0 Writing: 1 Waiting: 0"
            rw_parts = rw_line.split()
            reading = int(rw_parts[1])
            writing = int(rw_parts[3])
            waiting = int(rw_parts[5])
            
            return {
                "active_connections": active_connections,
                "accepts": accepts,
                "handled": handled,
                "requests": requests,
                "reading": reading,
                "writing": writing,
                "waiting": waiting
            }
            
        except Exception as e:
            logger.error(f"Failed to parse nginx status: {e}")
            return {"error": "Failed to parse status"}
    
    def test_config(self) -> bool:
        """Test nginx configuration validity."""
        try:
            nginx_container = self._get_nginx_container()
            result = nginx_container.exec_run(
                ["nginx", "-t"]
            )
            return result.exit_code == 0
        except Exception:
            return False
    
    def list_app_configs(self) -> List[str]:
        """List all app configurations managed by this manager."""
        try:
            configs = []
            for conf_file in self.conf_dir.glob("*.conf"):
                if conf_file.stem not in ["default", "nginx"]:  # Skip system configs
                    configs.append(conf_file.stem)
            return configs
        except Exception as e:
            logger.error(f"Failed to list app configs: {e}")
            return []

    def get_container_logs(self, lines: int = 100) -> str:
        """Get nginx container logs."""
        try:
            nginx_container = self._get_nginx_container()
            logs = nginx_container.logs(tail=lines, timestamps=True)
            return logs.decode('utf-8')
        except Exception as e:
            logger.error(f"Failed to get nginx logs: {e}")
            return f"Error getting logs: {e}"

    def restart_nginx(self) -> bool:
        """Restart the nginx container."""
        try:
            nginx_container = self._get_nginx_container()
            nginx_container.restart()
            logger.info(f"Restarted nginx container {self.nginx_container_name}")
            return True
        except Exception as e:
            logger.error(f"Failed to restart nginx container: {e}")
            return False

# For backward compatibility
NginxManager = DockerNginxManager

File: controller_go/nginx.go
package controller

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strings"
	"text/template"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
)

// NginxManagerImpl handles nginx configuration management and reloads.
type NginxManagerImpl struct {
	dockerClient  *client.Client
	containerName string
	confDir       string
	templatePath  string
	tmpl          *template.Template
}

// NewNginxManager creates a new manager from environment or params.
func NewNginxManager(containerName, confDir, templatePath string) (*NginxManagerImpl, error) {
	if containerName == "" {
		containerName = os.Getenv("ORCHESTRY_NGINX_CONTAINER")
	}
	if containerName == "" {
		return nil, fmt.Errorf("missing ORCHESTRY_NGINX_CONTAINER env var")
	}

	if confDir == "" {
		confDir = os.Getenv("ORCHESTRY_NGINX_CONF_DIR")
	}
	if confDir == "" {
		return nil, fmt.Errorf("missing ORCHESTRY_NGINX_CONF_DIR env var")
	}

	if templatePath == "" {
		templatePath = "configs/nginx_template_go.conf"
	}

	tmpl, err := template.ParseFiles(templatePath)
	if err != nil {
		return nil, fmt.Errorf("failed to load template: %w", err)
	}

	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return nil, fmt.Errorf("failed to create docker client: %w", err)
	}

	m := &NginxManagerImpl{
		dockerClient:  cli,
		containerName: containerName,
		confDir:       confDir,
		templatePath:  templatePath,
		tmpl:          tmpl,
	}

	if err := os.MkdirAll(confDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to ensure config dir: %w", err)
	}

	if err := m.ensureContainerRunning(); err != nil {
		return nil, err
	}

	log.Printf("NginxManager initialized for container %s", containerName)
	return m, nil
}

func (m *NginxManagerImpl) ensureContainerRunning() error {
	ctx := context.Background()
	container, err := m.dockerClient.ContainerInspect(ctx, m.containerName)
	if err != nil {
		return fmt.Errorf("nginx container %s not found: %w", m.containerName, err)
	}

	if !container.State.Running {
		log.Printf("Starting nginx container %s", m.containerName)
		if err := m.dockerClient.ContainerStart(ctx, m.containerName, types.ContainerStartOptions{}); err != nil {
			return fmt.Errorf("failed to start nginx container: %w", err)
		}
		log.Printf("Nginx container %s started", m.containerName)
	}
	return nil
}

func (m *NginxManagerImpl) validateAppName(name string) bool {
	if name == "" {
		return false
	}
	for _, r := range name {
		if !(r == '-' || r == '_' || (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9')) {
			return false
		}
	}
	return true
}

// UpdateUpstreams renders and applies an nginx config for an app.
func (m *NginxManagerImpl) UpdateUpstreams(app string, servers []Server) error {
	if !m.validateAppName(app) {
		return fmt.Errorf("invalid app name: %s", app)
	}
	if len(servers) == 0 {
		log.Printf("No servers provided for %s — removing config", app)
		return m.RemoveAppConfig(app)
	}

	confPath := filepath.Join(m.confDir, fmt.Sprintf("%s.conf", app))
	backupPath := confPath + ".backup"

	// Backup existing config
	if _, err := os.Stat(confPath); err == nil {
		if err := copyFile(confPath, backupPath); err != nil {
			return fmt.Errorf("failed to backup config: %w", err)
		}
	}

	// Convert servers to map format for template
	serverMaps := make([]map[string]interface{}, len(servers))
	for i, srv := range servers {
		serverMaps[i] = map[string]interface{}{
			"ip":   srv.IP,
			"port": srv.Port,
		}
	}

	// Render new config
	var buf bytes.Buffer
	err := m.tmpl.Execute(&buf, map[string]interface{}{
		"app":     app,
		"servers": serverMaps,
	})
	if err != nil {
		return fmt.Errorf("template render error: %w", err)
	}

	tmpFile := confPath + ".tmp"
	if err := os.WriteFile(tmpFile, buf.Bytes(), 0644); err != nil {
		return fmt.Errorf("failed to write temp config: %w", err)
	}
	os.Rename(tmpFile, confPath)

	// Test nginx config
	if !m.testConfig() {
		os.Remove(confPath)
		if _, err := os.Stat(backupPath); err == nil {
			os.Rename(backupPath, confPath)
		}
		return fmt.Errorf("nginx config test failed")
	}

	// Reload nginx
	if err := m.reloadNginx(); err != nil {
		os.Remove(confPath)
		if _, err := os.Stat(backupPath); err == nil {
			os.Rename(backupPath, confPath)
		}
		return fmt.Errorf("nginx reload failed: %w", err)
	}
	os.Remove(backupPath)
	log.Printf("Updated nginx config for %s", app)
	return nil
}

// RemoveAppConfig deletes the config file and reloads nginx.
func (m *NginxManagerImpl) RemoveAppConfig(app string) error {
	if !m.validateAppName(app) {
		return fmt.Errorf("invalid app name: %s", app)
	}
	confPath := filepath.Join(m.confDir, fmt.Sprintf("%s.conf", app))
	if err := os.Remove(confPath); err != nil && !os.IsNotExist(err) {
		return err
	}
	if !m.testConfig() {
		return fmt.Errorf("nginx config invalid after removal")
	}
	return m.reloadNginx()
}

// reloadNginx runs `nginx -s reload` in the container.
func (m *NginxManagerImpl) reloadNginx() error {
	return m.execInContainer([]string{"nginx", "-s", "reload"})
}

// testConfig runs `nginx -t` in the container.
func (m *NginxManagerImpl) testConfig() bool {
	err := m.execInContainer([]string{"nginx", "-t"})
	return err == nil
}

// execInContainer runs a command inside the nginx container.
func (m *NginxManagerImpl) execInContainer(cmd []string) error {
	ctx := context.Background()
	execID, err := m.dockerClient.ContainerExecCreate(ctx, m.containerName, types.ExecConfig{
		Cmd:          cmd,
		AttachStdout: true,
		AttachStderr: true,
	})
	if err != nil {
		return err
	}
	resp, err := m.dockerClient.ContainerExecAttach(ctx, execID.ID, types.ExecStartCheck{})
	if err != nil {
		return err
	}
	defer resp.Close()

	out, _ := io.ReadAll(resp.Reader)
	if !strings.Contains(string(out), "successful") && strings.Contains(strings.ToLower(string(out)), "error") {
		return fmt.Errorf("command failed: %s", string(out))
	}
	return nil
}

// GetNginxStatus fetches nginx status text and parses it.
func (m *NginxManagerImpl) GetNginxStatus() (map[string]interface{}, error) {
	ctx := context.Background()
	execID, err := m.dockerClient.ContainerExecCreate(ctx, m.containerName, types.ExecConfig{
		Cmd:          []string{"curl", "-s", "http://localhost:8080/nginx_status"},
		AttachStdout: true,
		AttachStderr: true,
	})
	if err != nil {
		return nil, err
	}
	resp, err := m.dockerClient.ContainerExecAttach(ctx, execID.ID, types.ExecStartCheck{})
	if err != nil {
		return nil, err
	}
	defer resp.Close()

	data, _ := io.ReadAll(resp.Reader)
	return parseNginxStatus(string(data))
}

// parseNginxStatus parses nginx stub_status text into metrics.
func parseNginxStatus(text string) (map[string]interface{}, error) {
	lines := strings.Split(strings.TrimSpace(text), "\n")
	if len(lines) < 4 {
		return nil, fmt.Errorf("unexpected nginx status output")
	}
	var active int
	fmt.Sscanf(lines[0], "Active connections: %d", &active)

	var accepts, handled, requests int
	fmt.Sscanf(lines[2], "%d %d %d", &accepts, &handled, &requests)

	var reading, writing, waiting int
	fmt.Sscanf(lines[3], "Reading: %d Writing: %d Waiting: %d", &reading, &writing, &waiting)

	return map[string]interface{}{
		"active_connections": active,
		"accepts":            accepts,
		"handled":            handled,
		"requests":           requests,
		"reading":            reading,
		"writing":            writing,
		"waiting":            waiting,
	}, nil
}

// GetContainerLogs returns nginx logs from Docker container.
func (m *NginxManagerImpl) GetContainerLogs(lines int) (string, error) {
	ctx := context.Background()
	opts := types.ContainerLogsOptions{
		ShowStdout: true,
		ShowStderr: true,
		Tail:       fmt.Sprint(lines),
		Timestamps: true,
	}
	rc, err := m.dockerClient.ContainerLogs(ctx, m.containerName, opts)
	if err != nil {
		return "", err
	}
	defer rc.Close()
	data, _ := io.ReadAll(rc)
	return string(data), nil
}

// RestartNginx restarts the nginx container.
func (m *NginxManagerImpl) RestartNginx() error {
	ctx := context.Background()
	timeout := 10
	return m.dockerClient.ContainerRestart(ctx, m.containerName, container.StopOptions{
		Timeout: &timeout,
	})
}

// Helper: copy file
func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()
	_, err = io.Copy(out, in)
	return err
}
